src/zkp/piprm.rs

```diff
@@ -6,141 +6,142 @@
 // License, Version 2.0 found in the LICENSE-APACHE file in the root directory
 // of this source tree.
 
-//! Implements the ZKP from Figure 17 of <https://eprint.iacr.org/2021/060.pdf>
+//! Implements a zero-knowledge proof that [`RingPedersen`] parameters were correctly
+//! constructed.
+//!
+//! In more detail, a valid [`RingPedersen`] object is compromised of a tuple `(N, s, t)`
+//! such that `s = t^λ mod N` for some secret `λ`. This module implements a zero-knowledge
+//! proof [`PiPrmProof`] of this fact. The proof is defined in Figure 17 of CGGMP[^cite].
+//!
+//! This proof utilizes the soundness parameter as specified [here](crate::parameters::SOUNDNESS_PARAMETER).
+//! In addition, it uses a standard Fiat-Shamir transformation to make the proof non-interactive.
+//!
+//! [^cite]: Ran Canetti, Rosario Gennaro, Steven Goldfeder, Nikolaos Makriyannis, and Udi Peled.
+//! UC Non-Interactive, Proactive, Threshold ECDSA with Identifiable Aborts.
+//! [EPrint archive, 2021](https://eprint.iacr.org/2021/060.pdf).
 
+use super::Proof;
 use crate::{errors::*, ring_pedersen::RingPedersen, utils::*};
 use libpaillier::unknown_order::BigNumber;
 use merlin::Transcript;
 use rand::{CryptoRng, RngCore};
 use serde::{Deserialize, Serialize};
 
-use super::Proof;
-
 // Soundness parameter lambda.
 const LAMBDA: usize = crate::parameters::SOUNDNESS_PARAMETER;
 
-/// Proof of ...
-///
-/// Each set of values must have length `LAMBDA`.
+/// Proof that given [`RingPedersen`] parameters are constructed correctly.
 #[derive(Debug, Clone, Serialize, Deserialize)]
 pub(crate) struct PiPrmProof {
-    a_values: Vec<BigNumber>,
-    e_values: Vec<u8>,
-    z_values: Vec<BigNumber>,
-}
-
-#[derive(Serialize)]
-pub(crate) struct PiPrmInput(RingPedersen);
-
-impl PiPrmInput {
-    pub(crate) fn new(ring_pedersen: RingPedersen) -> Self {
-        Self(ring_pedersen)
-    }
+    /// The public values computed by the prover (`A_i` in the paper).
+    public_values: Vec<BigNumber>,
```

Comment:
We could use `[BigNumber; LAMBDA]` directly, but this is not compatible with `serde` as of yet, although there is a [package](https://github.com/est31/serde-big-array) that supports it (which is sadly the [recommended approach](https://github.com/serde-rs/serde/pull/1860#pullrequestreview-735520420))

Reply:
It would be nice to encode the length explicitly but I'm okay to leave the `Vec` for now. That crate looks surprisingly un-popular for being such a basically useful function.

Reply:
Yeah, I'm surprised by that too...